src/lib/features/segment/segment-controller.ts

```diff
@@ -0,0 +1,470 @@
+            permission: NONE,
+            middleware: [
+                openApiService.validPath({
+                    summary: 'Get a segment',
+                    description: 'Retrieves a segment based on its ID.',
+                    tags: ['Segments'],
+                    operationId: 'getSegment',
+                    responses: {
+                        200: createResponseSchema('adminSegmentSchema'),
+                        ...getStandardResponses(404),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'post',
+            path: '',
+            handler: this.createSegment,
+            permission: CREATE_SEGMENT,
+            middleware: [
+                openApiService.validPath({
+                    summary: 'Create a new segment',
+                    description:
+                        'Creates a new segment using the payload provided',
+                    tags: ['Segments'],
+                    operationId: 'createSegment',
+                    requestBody: createRequestSchema('upsertSegmentSchema'),
+                    responses: {
+                        201: resourceCreatedResponseSchema(
+                            'adminSegmentSchema',
+                        ),
+                        ...getStandardResponses(400, 401, 403, 409, 415),
+                    },
+                }),
+            ],
+        });
+
+        this.route({
+            method: 'get',
+            path: '',
+            handler: this.getSegments,
+            permission: NONE,
+            middleware: [
+                openApiService.validPath({
+                    summary: 'Get all segments',
+                    description:
+                        'Retrieves all segments that exist in this Unleash instance.',
+                    tags: ['Segments'],
+                    operationId: 'getSegments',
+                    responses: {
+                        200: createResponseSchema('segmentsSchema'),
+                    },
+                }),
+            ],
+        });
+    }
+
+    async validate(
+        req: Request<unknown, unknown, { name: string }>,
+        res: Response,
+    ): Promise<void> {
+        const { name } = req.body;
+        await this.segmentService.validateName(name);
+        res.status(204).send();
+    }
+
+    async getSegmentsByStrategy(
+        req: Request<{ strategyId: string }>,
+        res: Response<SegmentsSchema>,
+    ): Promise<void> {
+        const { strategyId } = req.params;
+        const segments = await this.segmentService.getByStrategy(strategyId);
+
+        const responseBody = this.flagResolver.isEnabled('anonymiseEventLog')
+            ? {
+                  segments: anonymiseKeys(segments, ['createdBy']),
+              }
+            : { segments };
+
+        this.openApiService.respondWithValidation(
+            200,
+            res,
+            segmentsSchema.$id,
+            //@ts-ignore
```

Comment:
I do not understand why typescript complains here. Do you have any input  @thomasheartman?

Reply:
Hard to say without knowing what it's complaining about 🤷🏼 I can go check it out 🙋🏼

Reply:
Yeah, it looks like there's a schema issue here. It because of the declaration, it believes that the list of constraints within the segments can only ever be the empty array (`[]`).

Reply:
To be clear, the entire error message is:

```
Argument of type '{ segments: { id: number; name: string; description?: string | undefined; project?: string | undefined; constraints: { contextName: string; operator: "NOT_IN" | "IN" | "STR_ENDS_WITH" | "STR_STARTS_WITH" | ... 10 more ... | "SEMVER_LT"; values?: string[] | undefined; value?: string | undefined; inverted?: boolean | ...' is not assignable to parameter of type '{ [x: string]: unknown; segments?: { description?: string | null | undefined; usedInFeatures?: number | null | undefined; usedInProjects?: number | null | undefined; project?: string | null | undefined; ... 4 more ...; createdAt: string; }[] | undefined; }'.
   Types of property 'segments' are incompatible.
     Type '{ id: number; name: string; description?: string | undefined; project?: string | undefined; constraints: { contextName: string; operator: "NOT_IN" | "IN" | "STR_ENDS_WITH" | "STR_STARTS_WITH" | ... 10 more ... | "SEMVER_LT"; values?: string[] | undefined; value?: string | undefined; inverted?: boolean | undefined; c...' is not assignable to type '{ description?: string | null | undefined; usedInFeatures?: number | null | undefined; usedInProjects?: number | null | undefined; project?: string | null | undefined; createdBy?: string | null | undefined; id: number; name: string; constraints: []; createdAt: string; }[]'.
       Type '{ id: number; name: string; description?: string | undefined; project?: string | undefined; constraints: { contextName: string; operator: "NOT_IN" | "IN" | "STR_ENDS_WITH" | "STR_STARTS_WITH" | ... 10 more ... | "SEMVER_LT"; values?: string[] | undefined; value?: string | undefined; inverted?: boolean | undefined; c...' is not assignable to type '{ description?: string | null | undefined; usedInFeatures?: number | null | undefined; usedInProjects?: number | null | undefined; project?: string | null | undefined; createdBy?: string | null | undefined; id: number; name: string; constraints: []; createdAt: string; }'.
         Types of property 'constraints' are incompatible.
           Type '{ contextName: string; operator: "NOT_IN" | "IN" | "STR_ENDS_WITH" | "STR_STARTS_WITH" | "STR_CONTAINS" | "NUM_EQ" | "NUM_GT" | "NUM_GTE" | "NUM_LT" | "NUM_LTE" | "DATE_AFTER" | "DATE_BEFORE" | "SEMVER_EQ" | "SEMVER_GT" | "SEMVER_LT"; values?: string[] | undefined; value?: string | undefined; inverted?: boolean | un...' is not assignable to type '[]'.
             Target allows only 0 element(s) but source may have more. 
```

Reply:
So for instance, this'll give you the same error:

```ts
        const x: SegmentsSchema = {
            segments: [
                {
                    id: 7,
                    name: 'Test',
                    createdAt: '2023-09-14T08:34:12+02:00',
                    constraints: [{ type: 'IN', values: ['1', '2'] }],
                },
            ],
        };
```

Reply:
For a little more context: we've seen this a number of times, and it's usually due to a small error made somewhere that makes TS parse it as "must always be an empty list". Unfortunately, it doesn't appear to be the same thing every time (I think), so finding out what's going on is always a bit of a pain.  So in other words: I don't know exactly what is wrong here just yet.

Reply:
To fix this you should reference `constraintSchema` in components.schemas inside `segments-schema`. 

All used schemas should be referenced upwards, and since `adminSegmentSchema` references `constraintSchema`, it should also be referenced wherever `adminSegmentSchema` is referenced.

Reply:
Oh, damn, **you** just saved me a looooot of time if this works ❤️

Reply:
thanks guys. A bit surprised it worked in the enterprise. But good to have it resolved.